### PR TITLE
Add a shader option for fast math flags

### PR DESCRIPTION
**Pull request by Flakebi**
That allows setting them without using a command line option.
We can remove the command line option when the driver does not use it
anymore.

### CODE REVIEW

**Review by s-perron — APPROVED** [review]

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp [hunk 1]
```diff
@@ -971,6 +971,30 @@ FastMathFlags SPIRVToLLVM::getFastMathFlags(SPIRVValue *bv) {
 
   // For floating-point operations, if "FastMath" is enabled, set the "FastMath"
   // flags on the handled instruction
+  if (m_shaderOptions->fastMathFlags) {
+    if (m_shaderOptions->fastMathFlags & FastMathFlags::AllowReassoc)
+      fmf.setAllowReassoc();
```
Comment (kuhar) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:976]: nit: you can make this a little bit less redundant:

```c++
if (const flags_type flags  = m_shaderOptions->fastMathFlags) {
  if (flags & FastMathFlags::AllowReassoc)
     ...
     
```

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp [hunk 2]
```diff
@@ -971,6 +971,30 @@ FastMathFlags SPIRVToLLVM::getFastMathFlags(SPIRVValue *bv) {
...
+      fmf.setAllowReciprocal();
+
+    if (m_shaderOptions->fastMathFlags & FastMathFlags::AllowContract)
+      fmf.setAllowContract();
+
+    if (m_shaderOptions->fastMathFlags & FastMathFlags::ApproxFunc)
+      fmf.setApproxFunc();
+
+    return fmf;
+  }
```
Comment (kuhar) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:997]: This code is largely the same as for `SPIRVGenFastMathFlags`. Maybe it'd be better to factor this out into a helper function that takes fast math flags and returns `FastMathFlags`?

**Comment by amdvlk-admin** [comment]
### Test summary for commit 23adad890ec6514ea4f1bf5a98aa691af018fc1d
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev b638e90ca4e6e5a6fc4f00029d62f8e064aa18eb</li>
<li>SPVGEN: dev 34c9f9a74bc1b2b9a739933923920c5eeb9aa08f</li>
<li>XGL: dev 89a4cf115b9a2972ff00e2fe4959f4f1c8e7175b</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev afe13751c21d672a992d83f7d33856958f7d2f3b</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/199959)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 37595/66653 (56.4%)</li>
<li>Failed: 0/66653 (0.0%)</li>
<li>Not Supported: 29058/66653 (43.6%)</li><li>Warnings: 0/66653 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 37446/66653 (56.2%)</li>
<li>Failed: 0/66653 (0.0%)</li>
<li>Not Supported: 29207/66653 (43.8%)</li><li>Warnings: 0/66653 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 38653/66653 (58.0%)</li>
<li>Failed: 0/66653 (0.0%)</li>
<li>Not Supported: 28000/66653 (42.0%)</li><li>Warnings: 0/66653 (0.0%)</li>
</ul></details>


</ul></details>

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp (hunk 2 above)
Comment (Flakebi) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp:997]: The code using `SPIRVGenFastMathFlags` will go away soon, I just can’t do it in a single patch as it’s scattered across repositories.

Comment (kuhar) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp:997]: Ack.

File: include/vkgcDefs.h [hunk 3]
```diff
@@ -86,6 +86,7 @@
 //  %Version History
 //  | %Version | Change Description                                                                                    |
 //  | -------- | ----------------------------------------------------------------------------------------------------- |
+//  |     52.3 | Add fastMathFlags to PipelineShaderOptions                                                            |
```
Comment (amdrexu) [inline include/vkgcDefs.h:89]: I think you need to bump this: LLPC_INTERFACE_MINOR_VERSION

Comment (Flakebi) [reply include/vkgcDefs.h:89]: Right, thanks

**Comment by amdvlk-admin** [comment]
### Test summary for commit 36ea37f725072cff6a1af7790f965cf3be7ec67e
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev b638e90ca4e6e5a6fc4f00029d62f8e064aa18eb</li>
<li>SPVGEN: dev 34c9f9a74bc1b2b9a739933923920c5eeb9aa08f</li>
<li>XGL: dev 89a4cf115b9a2972ff00e2fe4959f4f1c8e7175b</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev afe13751c21d672a992d83f7d33856958f7d2f3b</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/199959)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 37595/66653 (56.4%)</li>
<li>Failed: 0/66653 (0.0%)</li>
<li>Not Supported: 29058/66653 (43.6%)</li><li>Warnings: 0/66653 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 37446/66653 (56.2%)</li>
<li>Failed: 0/66653 (0.0%)</li>
<li>Not Supported: 29207/66653 (43.8%)</li><li>Warnings: 0/66653 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 38653/66653 (58.0%)</li>
<li>Failed: 0/66653 (0.0%)</li>
<li>Not Supported: 28000/66653 (42.0%)</li><li>Warnings: 0/66653 (0.0%)</li>
</ul></details>


</ul></details>

**Review by kuhar — APPROVED** [review]

**Review by amdrexu — APPROVED** [review]

**Pull request merged**